internal/controller/keystoneapplicationcredential_controller.go

```diff
@@ -0,0 +1,543 @@
+	"github.com/gophercloud/gophercloud/v2/openstack/identity/v3/tokens"
+	keystonev1 "github.com/openstack-k8s-operators/keystone-operator/api/v1beta1"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	oko_secret "github.com/openstack-k8s-operators/lib-common/modules/common/secret"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/util"
+	corev1 "k8s.io/api/core/v1"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	ctrlLog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const acSecretFinalizer = "openstack.org/ac-secret-protection" // #nosec G101
+const finalizer = "openstack.org/applicationcredential"        // #nosec G101
+
+// ApplicationCredentialReconciler reconciles an ApplicationCredential object
+type ApplicationCredentialReconciler struct {
+	client.Client
+	Kclient       kubernetes.Interface
+	Log           logr.Logger
+	Scheme        *runtime.Scheme
+	EventRecorder record.EventRecorder
+}
+
+//+kubebuilder:rbac:groups=keystone.openstack.org,resources=keystoneapplicationcredentials,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=keystone.openstack.org,resources=keystoneapplicationcredentials/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=keystone.openstack.org,resources=keystoneapplicationcredentials/finalizers,verbs=update;patch
+//+kubebuilder:rbac:groups=core,resources=secrets/finalizers,verbs=get;list;create;update;delete;patch
+//+kubebuilder:rbac:groups=core,resources=events,verbs=create;patch
+
+// Reconcile reconciles a KeystoneApplicationCredential resource.
+func (r *ApplicationCredentialReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.GetLogger(ctx)
+
+	// Fetch the ApplicationCredential CR
+	instance := &keystonev1.KeystoneApplicationCredential{}
+	if err := r.Get(ctx, req.NamespacedName, instance); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Add finalizer only if not deleting and not already present
+	if instance.DeletionTimestamp.IsZero() && !controllerutil.ContainsFinalizer(instance, finalizer) {
+		base := instance.DeepCopy()
+		controllerutil.AddFinalizer(instance, finalizer)
+		if err := r.Patch(ctx, instance, client.MergeFrom(base)); err != nil {
+			if k8s_errors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Initialize conditions on first pass
+	if instance.Status.Conditions == nil {
```

Comment:
shouldn't we reset the conditions to init on each reconcile, like we do in other operators?